python/ray/data/datasource/file_based_datasource.py

```diff
@@ -209,15 +209,20 @@ def _resolve_paths_and_filesystem(
 
     resolved_paths = []
     for path in paths:
+        is_url = False
         if filesystem is not None:
             # If we provide a filesystem, _resolve_filesystem_and_path will not
             # slice off the protocol from the provided URI/path when resolved.
+            is_url = _is_url(path)
             path = _unwrap_protocol(path)
         resolved_filesystem, resolved_path = _resolve_filesystem_and_path(
             path, filesystem)
         if filesystem is None:
             filesystem = resolved_filesystem
         resolved_path = filesystem.normalize_path(resolved_path)
+        if is_url:
```

Comment:
What if there are other filesystems that actually handle url-encoding properly themselves? Should we whitelist specific filesystems like "s3"?

Reply:
AFAICT the current state is that no Arrow filesystems do URL-encoding, but fsspec filesystems might, and we need to guard against silent double-encoding if Arrow filesystems start doing automatic URL-encoding... So the options are:
1. URL-encode for every filesystem and have comprehensive filesystem test coverage to make sure that we aren't double-encoding.
2. Have a white list of Arrow filesystems that we URL-encode for, and have comprehensive filesystem test coverage to make sure that e.g. URLs with spaces are supported across all filesystems.
3. Put the onus on the user to encode their URLs, and tweak `_unwrap_protocol()` to be URL-encoding-aware (i.e. don't fail on the protocol prefix matching).

(3) seems to be the easiest to implement and maintain, but the worst UX, while both (1) and (2) will be more of a pain to maintain, but will offer better (Pandas-equivalent) UX. I'm leaning towards (2) right now.